### PR TITLE
Disallow non-string argument to `quote`

### LINKED ISSUE #2811: quote() function should reject non-string arguments
The `quote()` function in LibSass accepts arguments of any type. It should only accept string arguments, and produce errors for arguments of any other type.

https://github.com/sass/sass/issues/2597#issuecomment-451480529

### PR DESCRIPTION
**Pull request by glebm**
Tests: https://github.com/sass/sass-spec/pull/1380
Fixes #2811

### CODE REVIEW

File: src/fn_strings.cpp [hunk 1]
```diff
@@ -65,20 +65,12 @@ namespace Sass {
...
-      // only set quote mark to true if already a string
-      if (String_Quoted* qstr = Cast<String_Quoted>(arg)) {
-        qstr->quote_mark('*');
-        return qstr;
-      }
-      // all other nodes must be converted to a string node
-      std::string str(quote(arg->to_string(ctx.c_options), '"'));
-      String_Quoted* result = SASS_MEMORY_NEW(String_Quoted, pstate, str);
+      const String_Constant* s = ARG("$string", String_Constant);
+      String_Quoted* result = SASS_MEMORY_NEW(String_Quoted, pstate, s->value(), '"');
```
Comment (mgreter) [inline src/fn_strings.cpp]: AFAICS you are setting the quotemark to `"` on creation, but it is reset one line below to `*`.
IMO you can just use the following and remove the lines below:
`return SASS_MEMORY_NEW(String_Quoted, pstate, s->value(), '*');`

Comment (glebm) [reply src/fn_strings.cpp]: Do you know what the `*` quote mark means? Perhaps it should be `"`?

Comment (glebm) [reply src/fn_strings.cpp]: Turns out the `String_Quoted` constructor does a bunch of non-trivial things and passing it a quote mark doesn't actually set it.

Updated, PTAL.

Comment (mgreter) [reply src/fn_strings.cpp]: A * should mean that on output we choose the best quotemarks automatically.

**Comment by glebm** [comment]
Ah, should we change it to `*` then or always double-quote?

**Comment by mgreter** [comment]
Hehe, depends on what the expected output is, but basically it should be `*` IMO, but old
ruby sass sometimes preserved the original quotemark and sometimes didn't ...

**Comment by glebm** [comment]
I checked the output of Dart Sass and it also chooses a different quote depending on the input, so I've changed it to `*`.

**Comment by glebm** [comment]
Adding more tests as well

**Comment by glebm** [comment]
Added tests for quote detection to https://github.com/sass/sass-spec/pull/1380 and rebased. This should be good to merge.

**Pull request merged**